include/inl/gnc_utilities.inl

```diff
@@ -0,0 +1,138 @@
+  auto const qv = lin::ref<3, 1>(q, 0, 0);
+  res = v + lin::cross(
+      (static_cast<T>(2.0) * qv).eval(),
+      (lin::cross(qv, v) - q(3) * v).eval()
+    );
+}
+
+template <typename T>
+constexpr void rotate_frame(lin::Vector<T, 4> const &q, lin::Vector<T, 3> &v) {
+  rotate_frame(q, v, v);
+}
+
+// TODO : Potentially add diagonal references to lin
+template <typename T>
+inline void dcm_to_quat(lin::Matrix<T, 3, 3> const &M, lin::Vector<T, 4> &q) {
+  // Find the maximum among the diagonal elements and trace
+  T max = M(0, 0);
+  T trace = M(0, 0);
+  unsigned int idx = 0;
+  for (unsigned int i = 1; i < 3; i++) {
+    trace += M(i, i);
+    if (M(i, i) > max) {
+      max = M(i, i);
+      idx = i;
+    }
+  }
+  if (trace > max) {
+    max = trace;
+    idx = 3;
+  }
+
+  // Calculate the unnormalized quaternion
+  switch (idx) {
+    // Max was M(0, 0)
+    case 0:
+      q = {
+        static_cast<T>(2.0) * max - trace + static_cast<T>(1.0),
+        M(0, 1) + M(1, 0),
+        M(0, 2) + M(2, 0),
+        M(1, 2) - M(2, 1)
+      };
+      break;
+    // Max was M(1, 1)
+    case 1:
+      q = {
+        M(1, 0) + M(0, 1),
+        static_cast<T>(2.0) * max - trace + static_cast<T>(1.0),
+        M(1, 2) + M(2, 1),
+        M(2, 0) - M(0, 2)
+      };
+      break;
+    // Max was M(2, 2)
+    case 2:
+      q = {
+        M(2, 0) + M(0, 2),
+        M(2, 1) + M(1, 2),
+        static_cast<T>(2.0) * max - trace + static_cast<T>(1.0),
+        M(0, 1) - M(1, 0)
+      };
+      break;
+    // Max was trace(M)
+    default:
+      q = {
+        M(1, 2) - M(2, 1),
+        M(2, 0) - M(0, 2),
+        M(0, 1) - M(1, 0),
+        trace + static_cast<T>(1.0)
+      };
+  }
+
+  // Normalize the quaternion
+  q = q / lin::norm(q);
+}
+
+// TODO : Add more specific tolerancing to this function along with the QR
+// linear system solvers in lin/include/algorithms/*
+template <typename T>
+inline int triad(lin::Vector<T, 3> const &N_sun, lin::Vector<T, 3> const &N_mag,
+    lin::Vector<T, 3> const &B_sun, lin::Vector<T, 3> const &B_mag, lin::Vector<T, 4> &res) {
+  // Ensure that sun and magnetic field vectors aren't parallel or anitparallel
+  if (abs(lin::dot(N_sun, N_mag)) == static_cast<T>(1.0)) return 1;
+  if (abs(lin::dot(B_sun, B_mag)) == static_cast<T>(1.0)) return 2;
+
+  // Calculate right handed bases for the reference frame
+  lin::Vector<T, 3> const &n1 = N_sun;
+  lin::Vector<T, 3> const  n2 = lin::cross(n1, N_mag) / lin::norm(lin::cross(n1, N_mag));
+  lin::Vector<T, 3> const  n3 = lin::cross(n1, n2);
+
+  // Calculate right handed bases for the body frame
+  lin::Vector<T, 3> const &b1 = B_sun;
+  lin::Vector<T, 3> const  b2 = lin::cross(b1, B_mag) / lin::norm(lin::cross(b1, B_mag));
+  lin::Vector<T, 3> const  b3 = lin::cross(b1, b2);
+
+  // Calculate the DCM between the two frames and extract the quaternion
+  auto const Q = b1 * lin::transpose(n1) + b2 * lin::transpose(n2) + b3 * lin::transpose(n3);
```

Comment:
Just as another example, the auto keyword here actually represents a type like `Add<Multiply<Vector<>, Transpose<Vector<>>>, Add< ... >>` and the computation to calculate the result is done by `Q.eval()` on the next line.